pandas/types/concat.py

```diff
@@ -336,34 +347,63 @@ def convert_to_pydatetime(x, axis):
 
     # must be single dtype
     if len(typs) == 1:
+        _contains_datetime = any(typ.startswith('datetime') for typ in typs)
+        _contains_period = any(typ.startswith('period') for typ in typs)
 
-        if 'datetimetz' in typs:
-            # datetime with no tz should be stored as "datetime" in typs,
-            # thus no need to care
-
-            # we require ALL of the same tz for datetimetz
-            tzs = set([str(x.tz) for x in to_concat])
-            if len(tzs) == 1:
-                from pandas.tseries.index import DatetimeIndex
-                new_values = np.concatenate([x.tz_localize(None).asi8
-                                             for x in to_concat])
-                return DatetimeIndex(new_values, tz=list(tzs)[0])
+        if _contains_datetime:
 
-        elif 'datetime' in typs:
-            new_values = np.concatenate([x.view(np.int64) for x in to_concat],
-                                        axis=axis)
-            return new_values.view(_NS_DTYPE)
+            if 'datetime' in typs:
+                new_values = np.concatenate([x.view(np.int64) for x in
+                                             to_concat], axis=axis)
+                return new_values.view(_NS_DTYPE)
+            else:
+                # when to_concat has different tz, len(typs) > 1.
+                # thus no need to care
+                return _concat_datetimetz(to_concat)
 
         elif 'timedelta' in typs:
             new_values = np.concatenate([x.view(np.int64) for x in to_concat],
                                         axis=axis)
             return new_values.view(_TD_DTYPE)
 
+        elif _contains_period:
+            # PeriodIndex must be handled by PeriodIndex,
+            # Thus can't meet this condition ATM
+            # Must be changed when we adding PeriodDtype
+            raise NotImplementedError
+
     # need to coerce to object
     to_concat = [convert_to_pydatetime(x, axis) for x in to_concat]
     return np.concatenate(to_concat, axis=axis)
 
 
+def _concat_datetimetz(to_concat, name=None):
+    """
+    concat DatetimeIndex with the same tz
+    all inputs must be DatetimeIndex
+    it is used in DatetimeIndex.append also
+    """
+    # no need to localize because internal repr will not be changed
```

Comment:
I think we should do a run-time assert that these meet the guarantees (can do later)
e.g.
``assert len(set([ x.tz for x in to_concat ])) == 1``